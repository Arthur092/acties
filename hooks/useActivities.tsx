import { collection, getDoc, onSnapshot, query } from "firebase/firestore";
import React, { useState, useContext, createContext, useEffect } from "react";
import { ActivityType, RecordType } from "../constants/SampleData";
import { db, getActivityTypesByUser, getRecordsByUser } from "../firebase";
import { useAuth } from "./useAuth";

interface ContextValue {
    activityTypes: ActivityTypesState;
    records: RecordsState;
    getActivityTypes: Function;
    getRecords: Function
}

type ActivityTypesState = {
    data: Array<ActivityType>;
    isLoading: Boolean;
}

export type RecordsState = {
    data: Array<RecordType>;
    isLoading: Boolean;
}

const ActivitiesContext = createContext<ContextValue>({
    activityTypes: {
        data: [],
        isLoading: true
    },
    records: {
        data: [],
        isLoading: true
    },
    getActivityTypes: () => null,
    getRecords: () => null,
});

interface Props {
    children: JSX.Element
}

export function ProvideActivities({ children }: Props) {
    const activities: ContextValue = useProvideActivities();

    return (
        <ActivitiesContext.Provider value={activities}>
            { children }
        </ActivitiesContext.Provider>
    )
}

export const useActivities = () => useContext(ActivitiesContext);

function useProvideActivities(): ContextValue {
    const { user } = useAuth();
    const [newRecord, setNewRecord] = useState<RecordType | null>(null);

    useEffect(() => {
        const queryListener = query(collection(db, "Record"));
        const unsubscribe = onSnapshot(queryListener, (querySnapshot) => {
            querySnapshot.docChanges().forEach(change => {
                if (change.type === "added") {
                    const { activityType, date, quantity, userId } = change.doc.data();
                    const newRecord = {
                        id: change.doc.id,
                        activity: activityType,
                        date,
                        quantity,
                        userId
                    }
                    getDoc(newRecord.activity).then(activity => {
                        setNewRecord({
                            ...newRecord,
                            activity: activity.data() as ActivityType
                        });
                    });
                }
            })
        });

        return () => unsubscribe();
    },[]);

    useEffect(() => {
        if(newRecord){
            setRecords({
                ...records,
                data: [
                    ...records.data,
                    newRecord
                ],
            })
        }
    },[newRecord])

    const [activityTypes, setActivityTypes] = useState<ActivityTypesState>({
        data: [],
        isLoading: true
    });
    const [records, setRecords] = useState<RecordsState>({
        data: [],
        isLoading: true
    });

    const getActivityTypes = () => {
        if (user) {
            const { uid } = user;
            return getActivityTypesByUser(uid).then(newActivityTypes => {
                setActivityTypes({
                    ...activityTypes,
                    data: newActivityTypes,
                    isLoading: false
                })
            });
        }
    }

    const getRecords = () => {
        if (user) {
            const { uid } = user;
            return getRecordsByUser(uid).then(newRecords => {
                setRecords({
                    ...records,
                    data: newRecords,
                    isLoading: false
                })
            });
        }
    }

    return {
        activityTypes,
        records,
        getActivityTypes,
        getRecords
    }
}


