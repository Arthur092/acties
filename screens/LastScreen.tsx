import moment from 'moment';
import { StyleSheet } from 'react-native';
import { useEffect, useState } from 'react';
import { DataTable, List } from 'react-native-paper';
import { RecordType } from '../constants/SampleData';
import { useActivities } from '../hooks/useActivities';
import { Timestamp } from '@firebase/firestore-types'


// const optionsPerPage = [2, 3, 4];

const LastScreen = () => {
  // const [page, setPage] = useState<number>(0);
  // const [itemsPerPage, setItemsPerPage] = useState(optionsPerPage[0]);
  const [records, setRecords] = useState<Record<string, RecordType>>({});
  const { records: fetchRecords } = useActivities();

  useEffect(() => {
    const initialRecords: Record<string, RecordType> = fetchRecords.data.reduce((acc: Record<string, RecordType>, element: RecordType) => {
      const existingElement = Object.keys(acc).find(el => el === element.activity.name)
      if(existingElement){
        if((element.date as Timestamp).toMillis() > (acc[existingElement].date as Timestamp).toMillis()){
          acc = Object.assign(acc, {[existingElement]: element})
        }
      }else {
        acc = Object.assign(acc, {[element.activity.name]: element})
      }

      return acc;
    }, {});
    setRecords(initialRecords);
  }, [fetchRecords]);

  // useEffect(() => {
  //   setPage(0);
  // }, [itemsPerPage]);

  return (
    <DataTable>
      <DataTable.Header>
        <DataTable.Title style={styles.firstCell}>Activity</DataTable.Title>
        <DataTable.Title numeric>Number</DataTable.Title>
        <DataTable.Title numeric>date</DataTable.Title>
      </DataTable.Header>
      { Object.entries(records).map(([name, element], index) => (
        <DataTable.Row key={index}>
          <DataTable.Cell style={styles.firstCell}><List.Icon icon={element.activity.iconName} color={element.activity.iconColor}/> { name }</DataTable.Cell>
          <DataTable.Cell numeric>{ element.quantity }</DataTable.Cell>
          <DataTable.Cell numeric>{moment((element.date as Timestamp).toDate()).format('ll')}</DataTable.Cell>
        </DataTable.Row>
      ))}

      {/* <DataTable.Pagination
        page={page}
        numberOfPages={0}
        onPageChange={(page) => setPage(page)}
        // label="1-2 of 6"
        // numberOfItemsPerPage={itemsPerPage}
        // numberOfItemsPerPageList={[1,2,3]}
        // onItemsPerPageChange={setItemsPerPage}
        // selectPageDropdownLabel={'Rows per page'}
      /> */}
    </DataTable>
  );
}

const styles = StyleSheet.create({
  firstCell: {
    flex: 1.3,
  }
});

export default LastScreen;