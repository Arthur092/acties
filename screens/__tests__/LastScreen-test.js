import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react-native';
import LastScreen from '../LastScreen';
import { testActivityContextData } from '../../constants/mocks/activity';
import { ActivitiesContext } from '../../hooks/useActivities';

jest.useFakeTimers();

describe('Load LastScreen', () => {
  it('should renders correctly', () => {
    const tree = renderer
      .create(
        <ActivitiesContext.Provider value={testActivityContextData}>
          <LastScreen />
        </ActivitiesContext.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders with different value', () => {
    const { getByTestId } = render(
      <ActivitiesContext.Provider value={testActivityContextData}>
        <LastScreen />
      </ActivitiesContext.Provider>
    );
  });
});
