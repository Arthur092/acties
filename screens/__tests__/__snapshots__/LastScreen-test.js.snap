// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load LastScreen should renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "height": 48,
          "paddingHorizontal": 16,
        },
        Object {
          "borderBottomColor": "rgba(0, 0, 0, 0.12)",
        },
        undefined,
      ]
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignContent": "center",
            "flex": 1,
            "flexDirection": "row",
            "paddingVertical": 12,
          },
          undefined,
          Object {
            "flex": 1.3,
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "alignItems": "center",
                "fontSize": 12,
                "fontWeight": "500",
                "height": 24,
                "lineHeight": 24,
              },
              Object {
                "color": "rgba(0, 0, 0, 0.6)",
              },
            ],
          ]
        }
      >
        Activity
      </Text>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignContent": "center",
            "flex": 1,
            "flexDirection": "row",
            "paddingVertical": 12,
          },
          Object {
            "justifyContent": "flex-end",
          },
          undefined,
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "alignItems": "center",
                "fontSize": 12,
                "fontWeight": "500",
                "height": 24,
                "lineHeight": 24,
              },
              Object {
                "color": "rgba(0, 0, 0, 0.6)",
              },
            ],
          ]
        }
      >
        Number
      </Text>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignContent": "center",
            "flex": 1,
            "flexDirection": "row",
            "paddingVertical": 12,
          },
          Object {
            "justifyContent": "flex-end",
          },
          undefined,
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "alignItems": "center",
                "fontSize": 12,
                "fontWeight": "500",
                "height": 24,
                "lineHeight": 24,
              },
              Object {
                "color": "rgba(0, 0, 0, 0.6)",
              },
            ],
          ]
        }
      >
        date
      </Text>
    </View>
  </View>
</View>
`;
